Fix detection of local elasticsearch service

kevin-olbrich opened the pull request:
If elasticsearch is local and the URL was set via rake cli to "localhost" instead of "127.0.0.1", the service is not restarted.
This results in a broken state (search does not work, background tasks fail, etc.) if zammad is updated together with elasticsearch.

<!--
Hi there - a lot of love for starting a Pull Request 😍. Please ensure the following things before creation - thank you!

- Create your Pull Request against the develop branch. We don't accept changes to stable branches.
- Add a reference to the issue you are trying to fix. Just add a # and the number.
- Don't run `bundle update`. It's a honorable thought of you but some dependency versions (e.g. pg) are broken 😢 Just use `bundle install` if you introduce new dependencies instead.
- Run `rubocop` and `coffeelint` on your changes to respect the code style guide.
- Add tests for your changes. Feel free to ask for support. We are happy to help you.

* The upper textblock will be removed automatically when you submit your Pull Request *
-->

kevin-olbrich committed:
Fix detection of local elasticsearch service

If elasticsearch is local and the URL was set via rake cli to "localhost" instead of "127.0.0.1", the service is not restarted.
This results in a broken state (search does not work, background tasks fail, etc.) if zammad is updated together with elasticsearch.

monotek commented:
Not sure when this should happen?
The script explicitly sets 127.0.0.1.
Why should you want to change that? 

Usage of localhost could lead to problems with ipv6. 
See: https://github.com/zammad/zammad/issues/2365

thorsteneckel commented:
Hi @monotek - I'm with you that we should note (somewhere) that the usage of `localhost` is ambiguous and should be discouraged. Maybe the admin documentation would be the proper place - @MrGeneration ? 
@monotek - do you thing an additional check for `localhost` would be good to notify the user?
What about other local variations like e.g. `::1` etc.?

monotek commented:
If we want to support "::1" we should get sure that ipv6 is supported everywhere. I gueess that would need some testing...

I'd also like to know why the setting was changed to "localhost" in the first place to becom aware how likely this is for others or if it was changed by accident.

kevin-olbrich commented:
I changed the URL in the past to use an external cluster. This lead to many problems I did not want to further inspect and changed back to localhost. I did not know that it's using 127.0.0.1 by default.

While I understand your comment, the logic checks if the service is local. Local can be 127.0.0.1 as well as "localhost".

> Usage of localhost could lead to problems with ipv6.
> See: #2365

Actually "localhost" should resolve to both IPv4 and IPv6 with IPv6 preferred by the client. There are many reasons a connection can fail and the client retries with IPv4.
This is perfectly fine and is supposed to work exactly like that.

The script above checks if elasticsearch is local (thats that the codes says) but it lacks support for all possible values. Actually the upgrade is not done correctly in my setup, just because I entered "localhost" instead of 127.0.0.1.

Also, hardcoding IPv4 is bad practice as it's a legacy protocol with better alternatives like IPv6.
As ES5 is listening on both IPv4 and IPv6 by default, this should only cause problems on systems that are configures wrong.

kevin-olbrich commented:
> I'd also like to know why the setting was changed to "localhost" in the first place to becom aware how likely this is for others or if it was changed by accident.

I tried to use an external cluster which was no good idea. I used the same command to switch back to my local installation. And as there was no info that "localhost" should not be used, I did not ask.

At least for a clean install of CentOS 7, IPv6 is absolutely no problem. It just causes problems when something is not set up right like broken APs with radvd. Suchs problems should always be fixed by the net admins and not by the software. Disabling whole IPv6 via sysctl would also cause no problems, as localhost will only be looked up as 127.0.0.1.

monotek commented:
OK, i think we can add this check.
I just wanted to get sure that localhost was manually set.

kevin-olbrich commented:
Thank you very much!
